### PR TITLE
Improved compatibility with sklearn

### PR DESCRIPTION
**Pull request by daviddiazvico**
The original PR in keras-team/keras is [this](https://github.com/keras-team/keras/pull/12833
)

## Summary

Scikit-learn estimators are instanced only with hyper-parameter information. No knowledge on the data (X or y) is needed until the estimator is fitted. On the other hand, KerasRegressor and KerasClassifier receive as argument a build_fn that instantiates and compiles the estimator. At this point, the input and output shapes are needed to define the network's architecture. This is a subtle but important difference between Scikit-learn estimators and KerasClassifier/KerasRegressor.

Also, Scikit-learn's GridSearchCV and RandomizedSearchCV (and probably Scikit-optimize's BayesSearchCV) automatically infer the most appropriate CV folding (stratified or not) from the base_estimator. KerasClassifier and KerasRegressor, however, do not have the attribute needed for this to work.

## Related Issues

#32532

### CODE REVIEW

**Comment by daviddiazvico** [comment]
@googlebot I signed it!

**Comment by qlzh727** [comment]
Adding @fchollet for more input here.

I am not sure if the sklearn wrapper in keras is still under support or not, and I don't have much knowledge about it.

**Comment by daviddiazvico** [comment]
I've fixed a minor bug in the get_params method of BaseWrapper that I detected later. Just tell me if you want me to squash the commits.

**Comment by gbaned** [comment]
@daviddiazvico  Could you please resolve the conflicts? Thanks!

**Comment by gbaned** [comment]
@fchollet Can you please take a look on this PR? Thanks!

**Comment by daviddiazvico** [comment]
Can I do anything to speed this up? Thanks!

**Comment by adriangb** [comment]
I'm commenting here since I was working on a similar project and encountered another sklearn compatibility bug. @daviddiazvico would be happy to see if I can fix this for this PR. 

I found an issue using the `KerasClassifier` wrapper with any Keras Callback. I tested and confirmed on collab ([notebook](https://colab.research.google.com/drive/1IeTVO9R25R77vgI2G1ReU4rVGuW6bdVx)).

**Describe the current behavior**
Cross-validation (and presumably other sklearn functionality) is not compatible with `KerasClassifier` when used with a callback.

**Describe the expected behavior**
That all sklearn functions are applicable to the sklearn keras wrappers.

**Standalone code to reproduce the issue** 
https://colab.research.google.com/drive/1IeTVO9R25R77vgI2G1ReU4rVGuW6bdVx

**Other info / logs**
I believe the issue stems from the copying of `Callback` returning a new instance. The solution is probably to borrow some ideas from sklearn `Pipeline` or `Ensamble`, which are meta-estimators that deal with having nested estimator instances. I think `BaseWrapper` should be made some type of meta-estimator since it can contain other instantiated objects.

**Comment by qlzh727** [comment]
Ping @fchollet for review, and removing myself as reviewer.

**Comment by adriangb** [comment]
@fchollet @daviddiazvico I started working on this in a separate branch to evaluate the feasibility of making this close to the sklearn API by changing the implementation to mirror sklearn's `BaseEstimator`, `ClassifierMixin` and  `RegressorMixin`.

The source is on [my fork](https://github.com/adriangb/tensorflow/tree/wrapper-devel), a brief explanation below. It needs a lot of work on docs, cleaning up implementation, etc, but please take a look and let me know what you think, if it seems like a viable path forward I'll be happy to work on it some more.

The idea is similar to the current wrapper, but there was a lot of re-writing I had to do:
1. Incorporate most of the changes @daviddiazvico proposed but without breaking the existing API for `build_fn` (i.e. all of the existing tests still pass).
2. Moved the dictionary of __init__ parameters to self.__dict__ to comply with sklearn API.
3. Add `__getstate__` and `__setstate__` with logic for pickling keras `Model` and `Callback` instances.
4. Add `sample_weight` argument were needed for sklearn compatibility.
5. Create methods `_build_keras_model` and `_fit_keras_model` in `BaseWrapper` to clean up handling of the `X` and `y` parameters (previously, logic regarding categorical loss functions existed in `BaseWrapper`, now that functionality got moved to `KerasClassifier.fit`.
6. A lot of other minor things to fix new edge cases or existing incompatibilities.

This allows some cool things, for example, subclassing of `KerasClassifier` with hardcoded params:
```python3
class CustomClassifier(KerasClassifier):

    # define fit or model building params
    def __init___(self,
        hidden_layers=(100,),
        max_n_no_progress=10,
    ):
        self.hidden_layers=hidden_layers
        self.max_n_no_progress = max_n_no_progress
        # must manually assign each param
        # just like if you inherited from sklearn.base.BaseEstimator

        # can define callbacks anywhere as long as the attribute is called 'callbacks'
        self.callbacks = [EarlyStop(max_n_no_progress)]

    # define model building
    def __call__(X, y, n_classes_, hidden_layers):
        # can request X, y or any attribute of the class
        if X.shape[1] == 1:
          # easy to implement logic to dynamically pick last layer activation, etc.
        return model
```

Callbacks are also supported, and they can be pickled along with the rest of the class.

**Comment by fchollet** [comment]
Following our discussion at https://github.com/tensorflow/tensorflow/pull/37201, our current plan is to deprecate and no longer maintain the wrapper, and let the external Scikeras library take over. Thank you for your understanding.
